stix_shifter_modules/cybereason/stix_translation/query_constructor.py

```diff
@@ -0,0 +1,471 @@
+        :param qualifier: str, default in None
+        :return :None or str
+        """
+        if isinstance(expression, ComparisonExpression):  # Base Case
+            stix_object, stix_field = expression.object_path.split(':')
+            mapped_fields_array = self.dmm.map_field(stix_object, stix_field)
+            comparator = self.comparator_lookup[expression.comparator]
+            if expression.negated:
+                comparator = QueryStringPatternTranslator._format_negate(comparator)
+            value = self.__eval_comparison_value(expression, mapped_fields_array)
+            self._parse_mapped_fields(comparator, value, mapped_fields_array, qualifier)
+
+        elif isinstance(expression, CombinedComparisonExpression):
+            self._parse_expression(expression.expr1, qualifier)
+            self._parse_expression(expression.expr2, qualifier)
+        elif isinstance(expression, ObservationExpression):
+            self.__eval_observation_expression(expression, qualifier)
+
+        elif hasattr(expression, 'qualifier') and hasattr(expression, 'observation_expression'):
+            if isinstance(expression.observation_expression, CombinedObservationExpression):
+                self._parse_expression(expression.observation_expression.expr1, expression.qualifier)
+                self._parse_expression(expression.observation_expression.expr2, expression.qualifier)
+            else:
+                self._parse_expression(expression.observation_expression, expression.qualifier)
+
+        elif isinstance(expression, CombinedObservationExpression):
+            self._parse_expression(expression.expr1, qualifier)
+            self._parse_expression(expression.expr2, qualifier)
+        elif isinstance(expression, Pattern):
+            self._parse_expression(expression.expression)
+        else:
+            raise RuntimeError(f'Unknown Recursion Case for expression={expression}, '
+                               f'type(expression)={type(expression)}')
+
+    def parse_expression(self, pattern: Pattern):
+        if "ComparisonExpressionOperators.Or" in str(pattern) or "ObservationOperators.Or" in str(pattern):
+            raise NotImplementedError("OR operator is not supported in Cybereason")
+
+        self._parse_expression(pattern)
+
+
+def translate_pattern(pattern: Pattern, data_model_mapping, options):
+    """
+    Conversion of ANTLR pattern to Cybereason query
+    :param pattern: expression object, ANTLR parsed expression object
+    :param data_model_mapping: DataMapper object, mapping object obtained by parsing json
+    :param options: dict, time_range defaults to 5
+    :return: list, Cybereason queries
+    """
+    translated_query_strings = QueryStringPatternTranslator(pattern, data_model_mapping, options)
+    queries = translated_query_strings.qualified_queries
+    final_queries = [item for sublist in queries for item in sublist]
+    return final_queries
```

Comment:
why does one simple comparison translated into two of the exact same queries all the time? for example try translating this query- `"[ipv4-addr:value = '127.0.0.1'] START t'2021-10-01T11:00:00.000Z' STOP t'2021-10-07T11:00:00.003Z'"`

Reply:
Since Ipv4 stix element has 2 mappings( local address and remote address), two individual queries are generated.